tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -247,119 +244,197 @@ def _external_dep_name_from_bazel_dependency(bazel_dep):
+                _deduplicate_append(
+                    transitive_deps,
+                    bazel_rules[dep].get('_TRANSITIVE_DEPS', []))
+                _deduplicate_append(collapsed_deps,
+                                    bazel_rules[dep].get('_COLLAPSED_DEPS', []))
+                exclude_deps.update(bazel_rules[dep].get('_EXCLUDE_DEPS', []))
+
+        # This dep is a public target, add it as a dependency
+        if dep in bazel_label_to_dep_name:
+            _deduplicate_append(transitive_deps, [bazel_label_to_dep_name[dep]])
+            _deduplicate_append(collapsed_deps, [bazel_label_to_dep_name[dep]])
+            exclude_deps.update(bazel_rules[dep]['_TRANSITIVE_DEPS'])
+            exclude_deps.add(dep)
+            continue
 
-    target_name = target_dict['name']
-    bazel_deps = target_dict['_DEPS_BAZEL']
+        # This dep is an external target, add it as a dependency
+        if external_dep_name_maybe is not None:
+            _deduplicate_append(transitive_deps, [external_dep_name_maybe])
+            _deduplicate_append(collapsed_deps, [external_dep_name_maybe])
+            continue
 
-    # initial values
-    public_headers = set(target_dict['_PUBLIC_HEADERS_BAZEL'])
-    headers = set(target_dict['_HEADERS_BAZEL'])
-    src = set(target_dict['_SRC_BAZEL'])
-    deps = set()
+    # Direct dependencies are part of transitive dependencies
+    _deduplicate_append(transitive_deps, direct_deps)
+
+    # Removes all duplicated intermediate labels.
+    # This is the step that further shorten the deps list.
+    collapsed_deps = list(
+        filter(lambda x: x not in exclude_deps, collapsed_deps))
+
+    # Compute the final source files and headers for this build target whose
+    # name is `rule_name` (input argument of this function).
+    #
+    # Imaging a build target X has transitive deps [A, B, C, D, E]. C and E are
+    # public build targets. And D is a transitive dependency of C.
+    #
+    # Translate the condition into dependency graph:
+    #   X -> [A, B, C, D, E]
+    #   C -> [D]
+    #   Public targets [X, C, E]
+    #
+    # The collapsed dependencies of X: [C, E].
+    # The excluded dependencies of X: [C, D, E].
+    #
+    # Target X should include source files and headers of [X, A, B] as final
+    # build.
+    for dep in transitive_deps:
+        if dep not in exclude_deps:
+            if dep in bazel_rules:
+                collapsed_srcs.update(_extract_sources(bazel_rules[dep]))
+                collapsed_public_headers.update(
+                    _extract_public_headers(bazel_rules[dep]))
+                collapsed_headers.update(
+                    _extract_nonpublic_headers(bazel_rules[dep]))
+
+    # This item is a "visited" flag
+    bazel_rule['_PROCESSING_DONE'] = True
+    # Following items are described in the docstinrg.
+    bazel_rule['_TRANSITIVE_DEPS'] = transitive_deps
+    bazel_rule['_COLLAPSED_DEPS'] = list(sorted(collapsed_deps))
```

Comment:
if you're sorting the collapsed_deps anyway, you can save yourself the trouble  with storing it in a list and calling `_deduplicate_append`.

Reply:
Btw, since _COLLAPSED_DEPS get directly exposed as "deps" in build.yaml, you should know that the targets really need to come in  reverse topological order sorting.
In our Makefile, we simply pass the depdencies to the link in the order they are list in build.yaml and the order of passing libraries to the linker does matter. I think in your case this doesn't break anything since our Makefile currently only has a few targets and you happen to pass the deps in the right order (basically by coincidence), but that doesn't mean that the overall approach is correct (and not ensuring the ordering is correct could mean that the build suddenly breaks in the future because of e.g. a refactoring in BUILD where the generated deps don't come in the right order anyway).